### PR TITLE
Callable kernel

### LINKED ISSUE #86: Allow callable as kernel argument
For the built-in kernels you supply the data and the kernel you want to use, i.e.

```julia
model = svmtrain(Xtrain, ytrain, kernel=Kernel.Polynomial)

svmpredict(model, Xtest)
```

For a custom kernel you need to calculate the gram matrix manually:
```julia
K = make_gram(Xtrain, my_kernel)
model = svmtrain(K, ytrain, kernel=Kernel.Precomputed)

KK = make_gram(Xtrain, Xtest, my_kernel)
svmpredict(model, KK)
```

I think a more desirable way to do this would be to allow a callable as `kernel` argument (and calculate the gram matrix internally), i.e.
```julia
model = svmtrain(Xtrain, ytrain, kernel=my_kernel)

svmpredict(model, Xtest)
```
which is more syntactically consistent with the way the built-in kernels are used. Scikit-learn, for example, [allows this](https://scikit-learn.org/stable/modules/svm.html#custom-kernels) in addition to precomputed gram matrices.

### PR DESCRIPTION
**Pull request by till-m**
Resolves #86

### CODE REVIEW

File: test/runtests.jl
Comment (ablaom) [inline test/runtests.jl:290]: I suggest adding a comment explaining source of the reference result. (I'm guessing they come from using the built-in linear kernel, yes?)

File: src/LIBSVM.jl
Comment (ablaom) [inline src/LIBSVM.jl:110]: Discaimer: I'm not a maintainer of this package and comments are just suggestions.

Perhaps don't assume the callable object has type `Function`. I can construct objects that can be called like functions but they are not `Function`. So, if it's not one of the built-in types, I'd just assume it's callable. That's more generic. If  a user supplies callable, then they will get an informative error message telling them the object is not callable. 


In line 45 you then do `kernel::K` and make add `K` as a type parameter in line 43.

File: src/LIBSVM.jl
Comment (ablaom) [inline src/LIBSVM.jl:227]: My suggestion, is not to check kernel is valid function.

File: test/runtests.jl
Comment (ablaom) [reply test/runtests.jl:290]: Nice to have these other tests, by the way.

**Comment by ablaom** [comment]
@till-m Thanks for this.  

This looks good to me, my comment regarding use of `Function` not withstanding. You'll still need a maintainer to review this.

File: test/runtests.jl
Comment (till-m) [reply test/runtests.jl:290]: This test (and the `"Iris data"` test) are adapted from tests for the precomputed kernel. Maybe @barucden can pitch in on where the coefficients come from.

**Comment by till-m** [comment]
Thank you for your comments @ablaom, much appreciated! I'm currently waiting on the fate of #87 to be decided so that I can fix any merge conflicts that might arise, after which I will mark this PR ready for review.

**Comment by iblis17** [comment]
oh, well, I need some time to check PR #87. (quite busy this week

File: test/runtests.jl
Comment (barucden) [reply test/runtests.jl:290]: I am pretty sure I took it from https://github.com/JuliaML/LIBSVM.jl/pull/88/files

File: src/LIBSVM.jl
Comment (barucden) [inline src/LIBSVM.jl:331]: Is this not required anymore?

File: src/LIBSVM.jl
Comment (barucden) [inline src/LIBSVM.jl:248]: Would it be too bad to do
```julia
gram[i, j] = gram[j, i] = kernel_function(X[:, i], X[:, j])
```
?

**Comment by barucden** [comment]
I have a general question. Does this have any effect on serialization? 

TBH I have never serialized the `SVM` structure but I have always assumed that it would be possible. Can a `Function` be serialized?

File: src/LIBSVM.jl
Comment (till-m) [reply src/LIBSVM.jl:110]: I think it should be working, however if you can help me with constructing a non-`Function` kernel I can add a test case for it.

File: src/LIBSVM.jl
Comment (till-m) [reply src/LIBSVM.jl:248]: I would be fine with this. I'm pretty new to Julia, is there a standard way of handling a case like this?

File: src/LIBSVM.jl
Comment (barucden) [reply src/LIBSVM.jl:331]: Oh, I see.

The reason is the way libsvm handles precomputed kernels. See [this line](https://github.com/cjlin1/libsvm/blob/master/svm.cpp#L249) which simply assumes that `x[j][0]` contains the index of `j`-th instance. I am not sure why they decided to do it like that.

File: src/LIBSVM.jl
Comment (barucden) [reply src/LIBSVM.jl:248]: I guess we could consider this a standard way.

**Comment by till-m** [comment]
I would appreciate if someone could re-run the tests from `9ab1ef1`. At least the MacOS test seems to fail for unrelated reasons (`failed to resolve address for github.com: nodename nor servname provided, or not known`), not sure what the problem with the windows test is. I think I don't have permission to re-run jobs.

File: src/LIBSVM.jl
Comment (till-m) [reply src/LIBSVM.jl]: Added a test for non-`Function` callable.

File: src/LIBSVM.jl
Comment (till-m) [reply src/LIBSVM.jl:337]: I added your explanation as a comment in the code.

**Comment by till-m** [comment]
Thanks for re-running @iblis17, looks like the failures were a fluke. I'll mark this as ready for review now.

@barucden
> Can a `Function` be serialized?

it seems like this is indeed possible in principle:
```julia
using Serialization
foo(x) = x^2
serialize("foo.file", foo)
bar = deserialize("foo.file")
bar(2) # 4
``` 
I would guess that if one tries hard enough, the serialization will probably break down but hopefully such cases are very niche.

**Comment by barucden** [comment]
Unfortunately, it's not that easy. If you kill the Julia session after the serialization and try to load the file:
```julia
using Serialization
bar = deserialize("foo.file") #ERROR: UndefVarError: #foo not defined

foo(x) = x^2
bar = deserialize("foo.file")
bar(2) # 4
```

Essentially, the function can be deserialized only if it is defined within the session.

JLD.jl [deals](https://github.com/JuliaIO/JLD.jl#types-and-their-definitions) with a related issue (serializing custom structures) via `addrequire` function. IMHO, it would be sufficient to extend our README, document how to use the callable kernel, and mention that the kernel must be defined *before* deserialization.

**Comment by till-m** [comment]
Okay, that's... decidedly less useful.

> a test that covers the serialization/deserialization when a callable kernel is used

Like this, i.e. just to check whether it works _if_ the `kernel`-callable is defined? Or are you trying to cover something else?
```julia
distance(x) = x[1]^2 + x[2]^2

X = [0.72  0.68  0.28  0.75  0.47  0.26  0.95  0.0   0.95  0.39;
    0.49  0.07  0.67  0.94  0.4   0.98  0.21  0.29  0.91  0.16]

y = 0.5 .< [distance(x) for x in eachcol(X)]

kernel(x1, x2) = x1' * x2 + distance(x1) * distance(x2)

model  = svmtrain(X, y, kernel=kernel)
serialize("foo.file", model)


model₂ = deserialize("foo.file")
T = [0.57  0.56  0.57  0.51;
    0.9   0.37  0.04  0.76]

ŷ = 0.5 .< [distance(x) for x in eachcol(T)]
ỹ, _ = svmpredict(model₂, T)
@test ŷ == ỹ
```

**Comment by barucden** [comment]
Yes, I think it could be even simpler:

```julia
distance(x) = x[1]^2 + x[2]^2

X = [0.72  0.68  0.28  0.75  0.47  0.26  0.95  0.0   0.95  0.39;
     0.49  0.07  0.67  0.94  0.4   0.98  0.21  0.29  0.91  0.16]

y = 0.5 .< [distance(x) for x in eachcol(X)]

kernel(x1, x2) = x1' * x2 + distance(x1) * distance(x2)

model₁ = svmtrain(X, y, kernel=kernel)
serialize("foo.file", model)

model₂ = deserialize("foo.file")

y₁ = svmpredict(model₁, X)
y₂ = svmpredict(model₂, X)

@test y₁ == y₂
```

Just to check that the loaded model works (and works the same).

File: src/LIBSVM.jl
Comment (iblis17) [inline src/LIBSVM.jl:146]: I think we can simply have this?
```suggestion
    if kernel == Int32(Kernel.Precomputed)
```

File: src/LIBSVM.jl
Comment (iblis17) [inline src/LIBSVM.jl:240]: Doubling assignment on `i = j` is fine for me.
```suggestion
                gram[i, j] = gram[j, i] = kernel_function(X[:, i], X[:, j])
```

File: src/LIBSVM.jl
Comment (iblis17) [inline src/LIBSVM.jl:235]:
```suggestion
    for i ∈ 1:ntrain, j ∈ 1:i
```

File: src/LIBSVM.jl
Comment (iblis17) [inline src/LIBSVM.jl:256]: Can we assume that user defined `kernel` function doesn't modify the input?
```suggestion
    for (i, idx) in enumerate(SVs.indices)
        x = @view SVs.X[:, i]
        for j ∈ 1:npredict
            gram[idx, j] = kernel(x, @view(T[:, j]))
        end
    end
```

File: src/LIBSVM.jl
Comment (iblis17) [inline src/LIBSVM.jl:382]:
```suggestion
    nodes, nodeptrs = if !(kernel isa Kernel.KERNEL)
        gram2nodes(data2gram(kernel, X))
    elseif kernel == Kernel.Precomputed
        gram2nodes(X)
    else
        instances2nodes(X)
    end
```

File: test/runtests.jl
Comment (iblis17) [inline test/runtests.jl:379]:
```suggestion
             0.49  0.07  0.67  0.94  0.4   0.98  0.21  0.29  0.91  0.16]
```

File: test/runtests.jl
Comment (iblis17) [inline test/runtests.jl:389]:
```suggestion
             0.9   0.37  0.04  0.76]
```

File: src/LIBSVM.jl
Comment (iblis17) [inline src/LIBSVM.jl:133]:
```suggestion
    kernel = Int32(ifelse(mod.kernel isa Kernel.KERNEL, mod.kernel, Kernel.Precomputed))
```

File: src/LIBSVM.jl
Comment (iblis17) [inline src/LIBSVM.jl:345]:
```suggestion
    _kernel = Int32(ifelse(mod.kernel isa Kernel.KERNEL, kernel, Kernel.Precomputed))
```

File: src/LIBSVM.jl
Comment (till-m) [reply src/LIBSVM.jl:256]: With `@view` you make it modifiable, right? I cannot imagine a usecase where one would like the kernel to modify it's inputs, the kernel should be a similarity measure and as such a pure function. Otherwise you could run into all kinds of problems like the order of evaluation being important, the kernel evaluating to different values for a datapoint both present in the training and testing dataset, etc.

File: src/LIBSVM.jl
Comment (till-m) [reply src/LIBSVM.jl:235]: I notice that all
```julia
for i in 1:n
for i = 1:n
for i ∈ 1:n
```
are options for iterator assignment in Julia. Are they all functionally and semantically the same? If yes I'd suggest settling on one version and I'd make sure that at least all the for-loops that are modified/added in this PR use that version.

Comment (iblis17) [reply src/LIBSVM.jl:235]: yes, all of them are the same. I'm okay with any one. `∈` is just my personal preference.

File: src/LIBSVM.jl
Comment (iblis17) [reply src/LIBSVM.jl:256]: yes, `@view` doesn't copy the data to a new array, and eliminate the memory allocation which is time consuming.

So, I think using `@view` here is reasonable.

Comment (till-m) [reply src/LIBSVM.jl:256]: Ah I think I might've misunderstood your intention, I was thinking you were enabling modification intentionally.

I think it is very unlikely that someone modifies the input, but possible. I don't know how large the performance benefit is but if you think it's worth it, that's fine with me. Maybe we just state that the `kernel` should not modify the input in the README/documentation?

File: src/LIBSVM.jl
Comment (till-m) [reply src/LIBSVM.jl:382]: Would 
```julia
    if !(kernel isa Kernel.KERNEL)
        nodes, nodeptrs = gram2nodes(data2gram(kernel, X))
    elseif kernel == Kernel.Precomputed
        nodes, nodeptrs = gram2nodes(X)
    else
       nodes, nodeptrs =  instances2nodes(X)
    end
```
be acceptable to you? I think it's a bit easier to read & doesn't take any more lines but otherwise I don't feel strongly about it.

Sidenote: I didn't know about the `kernel isa Kernel.KERNEL` syntax, that seems more elegant than `isa(kernel, Kernel.KERNEL)`. In any case, I would also try to be consistent within the PR but I'd be fine with either style, really.

Comment (iblis17) [reply src/LIBSVM.jl:382]: ha, :ok_hand:

File: src/LIBSVM.jl
Comment (iblis17) [reply src/LIBSVM.jl:256]: > Maybe we just state that the kernel should not modify the input in the README/documentation?

ah, it's fine for me.

**Comment by till-m** [comment]
The windows julia 1.3 test failed again in `2e8c64e`, not sure why. Other than that I think I got all of the changes and it seems to be working.

File: src/LIBSVM.jl
Comment (iblis17) [inline src/LIBSVM.jl:310]: Could you update the docstring as well? Just note that the `kernel` argument could be `Base.Callable`.

Comment (till-m) [reply src/LIBSVM.jl:310]: Updated. Let me know if it works for you.

**Comment by iblis17** [comment]
ah, LGTM. :+1: 

I'm going to merge this PR tmr if no objection.

**Comment by till-m** [comment]
I think one last question to settle would be whether the `ScikitLearn.jl` API should also be supported. I assume that requires only the relaxation of a few type checks.

**Comment by iblis17** [comment]
> I think one last question to settle would be whether the ScikitLearn.jl API should also be supported. I assume that requires only the relaxation of a few type checks.

Yeah, it should also be supported. You want to include it in this PR or add another PR?

**Comment by till-m** [comment]
> You want to include it in this PR or add another PR?

Whichever is more standard or whichever you prefer.

**Comment by iblis17** [comment]
> Whichever is more standard or whichever you prefer.

While you start to modify the code,
if you find that the changes are complex and maybe need more than 3 commits to clarify them,
adding a new PR for them will be a better option.
Otherwise, including it in the original PR is fine.

**Comment by till-m** [comment]
I removed the type checks and added a test case. Seems to be working.

**Review by iblis17 — APPROVED** [review]

**Comment by iblis17** [comment]
Out of my curiosity, is there a kernel function such that the Gram matrix isn't symmetric (`Gram[i, j] != Gram[j, i]`)?

**Comment by till-m** [comment]
Kernels are associated with an inner product of some Hilbert space by a feature map φ that maps to that Hilbert space: k(x1, x2) = <φ(x1), φ(x2)> and inner products are (hermetian) symmetric. Since a symmetric kernel function would lead to a symmetric Gram matrix, all _real_ kernels produce Gram matrices that are symmetric.
Now I have never seen a complex kernel function (in an ML context) and I don't know if they exist (or if LIBSVM supports complex valued precomputed gram matrices). I've checked a few textbooks (_Mathematics for Machine Learning_ by Deisenroth, _Support Vector Machines_ by Steinwart and Schölkopf/Smola's _Learning with Kernels_) and they all explicitly mention that the Gram matrix is symmetric and/or define kernels as real-valued functions.

**Comment by iblis17** [comment]
Tanks for your contributions!

**Comment by iblis17** [comment]
I will release these patches with version = 0.8.0 tmr.

**Comment by iblis17** [comment]
> Kernels are associated with an inner product of some Hilbert space by a feature map φ that maps to that Hilbert space: k(x1, x2) = <φ(x1), φ(x2)> and inner products are (hermetian) symmetric. Since a symmetric kernel function would lead to a symmetric Gram matrix, all real kernels produce Gram matrices that are symmetric. ...

A side note: I just found a work [1]. IIUC, if user can provide a positive definite matrix, we can safely pass it to libsvm.

[1] Tsuda, K. (1999). Support vector classifier with asymmetric kernel functions. In in European Symposium on Artificial Neural Networks (ESANN.

**Comment by till-m** [comment]
Thanks for the guidance @ablaom @barucden @iblis17.

**Pull request merged**
